lib/src/rewrite.rs

```diff
@@ -1180,3 +1180,82 @@ pub fn squash_commits<'repo>(
         abandoned_commits,
     }))
 }
+
+/// Find divergent commits from the target that are already present with
+/// identical contents in the destination. These commits should be able to be
+/// safely abandoned.
+pub fn find_duplicate_divergent_commits(
+    repo: &dyn Repo,
+    new_parent_ids: &[CommitId],
+    target: &MoveCommitsTarget,
+) -> BackendResult<Vec<Commit>> {
+    let rebased_commits: Vec<Commit> = match target {
+        MoveCommitsTarget::Roots(root_ids) => RevsetExpression::commits(root_ids.clone())
+            .descendants()
+            .evaluate(repo)
+            .map_err(|err| err.into_backend_error())?
+            .iter()
+            .commits(repo.store())
+            .try_collect()
+            .map_err(|err| err.into_backend_error())?,
+        MoveCommitsTarget::Commits(commit_ids) => commit_ids
+            .iter()
+            .map(|commit_id| repo.store().get_commit(commit_id))
+            .try_collect()?,
+    };
+    let rebased_commit_ids: HashSet<&CommitId> = rebased_commits.iter().map(Commit::id).collect();
+
+    // For each divergent change being rebased, we want to find all of the other
+    // commits with the same change ID which are not being rebased.
+    let divergent_changes: Vec<(&Commit, Vec<CommitId>)> = rebased_commits
+        .iter()
+        .map(|rebased_commit| {
+            let mut ancestor_candidates = repo
+                .resolve_change_id(rebased_commit.change_id())
+                .unwrap_or_default();
+            ancestor_candidates.retain(|commit_id| !rebased_commit_ids.contains(commit_id));
+            (rebased_commit, ancestor_candidates)
+        })
+        .filter(|(_, candidates)| !candidates.is_empty())
+        .collect();
+    if divergent_changes.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let rebase_root_commit_ids = match target {
+        MoveCommitsTarget::Roots(root_ids) => root_ids,
+        MoveCommitsTarget::Commits(commit_ids) => commit_ids,
+    };
+
+    // We only care about divergent changes which are new ancestors of the rebased
+    // commits, not ones which were already ancestors of the rebased commits.
+    let is_new_ancestor = RevsetExpression::commits(rebase_root_commit_ids.clone())
+        .range(&RevsetExpression::commits(new_parent_ids.to_owned()))
+        .evaluate(repo)
+        .map_err(|err| err.into_backend_error())?
+        .containing_fn();
+
+    let mut duplicate_divergent = Vec::new();
+    // Checking every pair of commits between these two sets could be expensive if
+    // there are several commits with the same change ID. However, it should be
+    // uncommon to have more than a couple commits with the same change ID being
+    // rebased at the same time, so it should be good enough in practice.
+    for (rebased_commit, ancestor_candidates) in divergent_changes {
```

Comment:
nit: `rebased_commit` in this context might be confusing because it is the commit "to be" rebased. Maybe we can use `target_commits`, `target_roots`, etc. as we call that way in `move_commits()`?